charlotte_core/src/arch/x86_64/mod.rs

```diff
@@ -231,4 +270,119 @@ impl Api {
         logln!("Contiguous frame allocation and deallocation test complete.");
         logln!("Physical Memory Manager test suite finished.");
     }
+
+    fn vmm_self_test() {
+        logln!("Beginning VMM Self Test...");
+        let cr3 = unsafe { asm_get_cr3() };
+        let mut pm = PageMap::from_cr3(cr3).expect("Failed to create PageMap from CR3 value.");
+        logln!("PageMap created from current CR3 value.");
+
+        logln!("Starting page mapping test...");
+        let frame = PHYSICAL_FRAME_ALLOCATOR
+            .lock()
+            .allocate()
+            .expect("Failed to allocate frame.");
+        let vaddr = match VirtualAddress::try_from(0xFFFF800000000000) {
+            Ok(vaddr) => vaddr,
+            Err(e) => {
+                panic!("Failed to create VirtualAddress: {:?}", e);
+            }
+        };
+        pm.map_page(
+            vaddr,
+            frame,
+            PteFlags::Write as u64 | PteFlags::Global as u64 | PteFlags::NoExecute as u64,
+        );
+        logln!(
+            "Mapped page at virtual address: {:?} to physical frame: {:?}",
+            vaddr,
+            frame
+        );
+        unsafe {
+            let ptr = <*mut u64>::from(vaddr);
+            ptr.write(0xdeadbeef);
+            logln!("Wrote 0xdeadbeef to virtual address: {:?}", vaddr);
+            let val = ptr.read();
+            logln!("Read value: 0x{:x} from virtual address: {:?}", val, vaddr);
+        }
+        let _ = pm.unmap_page(vaddr);
+        logln!("Unmapped page at virtual address: {:?}", vaddr);
+        let _ = PHYSICAL_FRAME_ALLOCATOR.lock().deallocate(frame);
+        logln!("Deallocated frame: {:?}", frame);
+        logln!("Page mapping test successful.");
+
+        logln!("Starting large page mapping test...");
+        let large_frame = PHYSICAL_FRAME_ALLOCATOR
+            .lock()
+            .allocate_contiguous(512, 4096 * 512)
+            .expect("Failed to allocate frames.");
+        let vaddr = match VirtualAddress::try_from(0xFFFF800000000000) {
+            Ok(vaddr) => vaddr,
+            Err(e) => {
+                panic!("Failed to create VirtualAddress: {:?}", e);
+            }
+        };
+        pm.map_large_page(
+            vaddr,
+            large_frame,
+            PteFlags::Write as u64 | PteFlags::Global as u64 | PteFlags::NoExecute as u64,
+        );
```

Comment:
Please ignore or include `let _ =`  to suppress the returned error. Ideally, as this is testing the VMM, handeling the error and writing out a log accordingly would be even better.

Reply:
+1